echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/cdevents/CDEventsBuilderService.java

```diff
@@ -0,0 +1,213 @@
+            .map(c -> (String) c.get("link"))
+            .orElseThrow(FieldNotFoundException::new);
+
+    String executionId =
+        Optional.ofNullable(event.content)
+            .map(e -> (Map) e.get("execution"))
+            .map(e -> (String) e.get("id"))
+            .orElse(null);
+
+    String executionUrl =
+        String.format(
+            "%s/#/applications/%s/%s/%s",
+            spinnakerUrl,
+            application,
+            configType == "stage" ? "executions/details" : configLink,
+            executionId);
+
+    String executionName =
+        Optional.ofNullable(event.content)
+            .map(e -> (Map) e.get("execution"))
+            .map(e -> (String) e.get("name"))
+            .orElse(null);
+
+    String cdEventsType =
+        Optional.ofNullable(preference)
+            .map(p -> (String) p.get("cdEventsType"))
+            .orElseThrow(FieldNotFoundException::new);
+
+    CloudEvent ceToSend =
+        buildCloudEventWithCDEventType(
+            cdEventsType, executionId, executionUrl, executionName, spinnakerUrl, status);
+    if (ceToSend == null) {
+      log.error("Failed to created CDEvent with type {} as CloudEvent", cdEventsType);
+      throw new CDEventsException("Failed to created CDEvent as CloudEvent");
+    }
+    return ceToSend;
+  }
+
+  private CloudEvent buildCloudEventWithCDEventType(
+      String cdEventsType,
+      String executionId,
+      String executionUrl,
+      String executionName,
+      String spinnakerUrl,
+      String status) {
+    CloudEvent ceToSend = null;
+    switch (cdEventsType) {
+      case "dev.cdevents.pipelinerun.queued":
+        ceToSend =
+            createPipelineRunQueuedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.pipelinerun.started":
+        ceToSend =
+            createPipelineRunStartedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.pipelinerun.finished":
+        ceToSend =
+            createPipelineRunFinishedEvent(
+                executionId, executionUrl, executionName, spinnakerUrl, status);
+        break;
+      case "dev.cdevents.taskrun.started":
+        ceToSend =
+            createTaskRunStartedEvent(executionId, executionUrl, executionName, spinnakerUrl);
+        break;
+      case "dev.cdevents.taskrun.finished":
+        ceToSend =
+            createTaskRunFinishedEvent(
+                executionId, executionUrl, executionName, spinnakerUrl, status);
+        break;
+      default:
+        throw new CDEventsException(
+            "Invalid CDEvents Type " + cdEventsType + " provided to create CDEvent");
+    }
+    return ceToSend;
+  }
+
+  private CloudEvent createTaskRunFinishedEvent(
+      String executionId,
+      String executionUrl,
+      String executionName,
+      String spinnakerUrl,
+      String status) {
+    TaskRunFinishedCDEvent cdEvent = new TaskRunFinishedCDEvent();
```

Comment:
Then all these classes can just implement that interface I mentioned [above](https://github.com/spinnaker/echo/pull/1295/files/e7cf71404335dfe18a6391b54b0449006d33213e#r1359232738)